File: packages/discord/src/commands/unsubscribe.ts
import { PREFIX } from "@daily-news-discord/environment";
import { Command } from "../utils/Command";
import {
  deleteGuildChannel,
  deleteGuildChannels,
  getGuildChannel,
  getGuildChannels,
} from "@daily-news-discord/database";
import { categoryArray } from "@daily-news-discord/news";

export default new Command({
  name: "unsubscribe",
  description: "Unsubscribe News to this channel",
  aliases: ["unsub"],
  usage: `${PREFIX}unsubscribe <category>`,
  execute: async (msg, args) => {
    if (!msg.guild) {
      msg.reply("Please use this command in a server");
      return;
    }

    const category = args[0];
    if (!category) {
      msg.reply("Please specify a category");
      return;
    }

    // Check if category is valid
    if (!categoryArray.includes(category) && category !== "all") {
      msg.reply("Please specify a valid category");
      return;
    }

    const channel = msg.channel;
    const guild = msg.guild;

    if (category === "all") {
      const guildChannels = await getGuildChannels(guild.id, channel.id);
      if (guildChannels.length === 0) {
        msg.reply("You are not subscribed to any categories");
        return;
      }

      await deleteGuildChannels(guild.id, channel.id);
      msg.reply("You are now unsubscribed from all categories");
      return;
    }

    const guildChannel = await getGuildChannel(guild.id, channel.id, category);
    if (!guildChannel) {
      msg.reply("You are already unsubscribed to this channel");
      return;
    }

    await deleteGuildChannel(guild.id, channel.id, category);
    msg.reply("You are now unsubscribed to this channel");
  },
});

File: commands/unsubscribe.go
package commands

import (
	"log"
	"strings"

	"github.com/SushiWaUmai/daily-news-discord/db"
	"github.com/SushiWaUmai/daily-news-discord/news"
	"github.com/bwmarrin/discordgo"
)

func init() {
	var categoryChoices []*discordgo.ApplicationCommandOptionChoice
	for _, c := range news.Categories {
		categoryChoices = append(categoryChoices, &discordgo.ApplicationCommandOptionChoice{
			Name:  c,
			Value: c,
		})
	}
	categoryChoices = append(categoryChoices, &discordgo.ApplicationCommandOptionChoice{
		Name:  "all",
		Value: "all",
	})

	var defaultMemberPermissions int64 = discordgo.PermissionManageServer
	dmPermission := false

	createCommand(&discordgo.ApplicationCommand{
		Name:        "unsubscribe",
		Description: "Unsubscribe News to this channel",
		DefaultMemberPermissions: &defaultMemberPermissions,
		DMPermission:             &dmPermission,
		Options: []*discordgo.ApplicationCommandOption{
			{
				Type:        discordgo.ApplicationCommandOptionString,
				Name:        "category",
				Description: "The news category you want to unsubscribe from",
				Required:    true,
				Choices:     categoryChoices,
			},
		},
	}, func(dg *discordgo.Session, i *discordgo.InteractionCreate) {
		guildID := i.GuildID
		channelID := i.ChannelID
		category := strings.ToLower(i.ApplicationCommandData().Options[0].StringValue())

		if category == "all" {
			db.DeleteGuildChannels(guildID, channelID)

			dg.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
				Type: discordgo.InteractionResponseChannelMessageWithSource,
				Data: &discordgo.InteractionResponseData{
					Content: "You are now unsubscribed from all categories",
				},
			})
			return
		}

		guildChannel, _ := db.GetGuildChannel(guildID, channelID, category)
		if guildChannel == nil {
			dg.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
				Type: discordgo.InteractionResponseChannelMessageWithSource,
				Data: &discordgo.InteractionResponseData{
					Content: "You are already unsubscribed to this channel",
				},
			})
			return
		}

		err := db.DeleteGuildChannel(guildID, channelID, category)
		if err != nil {
			log.Println("Failed to delete GuildChannel", err)
			return
		}

		dg.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
			Type: discordgo.InteractionResponseChannelMessageWithSource,
			Data: &discordgo.InteractionResponseData{
				Content: "You are now unsubscribed to this channel",
			},
		})
	})
}
